### PR TITLE
Set system-owned envs after user-provided envs

### PR DESCRIPTION
**Pull request by kevin-bates**
The changes introduced in #1668 to indicate the runtime name within an operation failed when an operation referenced `ELYRA_RUNTIME_ENV` on KFP and Local runtimes (Airflow was fine).  This corrects those issues and addresses test cases to better handle that scenario - since that _is_ the use case that is being addressed.

### How was this pull request tested?
Extended the tests to include cases where the system-owned variables are set in the operation to ensure the system values are properly reflected.  Ran a test against KFP since the unit tests test this functionality at a lower level for the Kfp and Airflow processors.  (The local test can address this since it includes an end-to-end operation.)

Fixes: #1696

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by lresende** [comment]
Should we allow the user to discover + set these at all?

**Comment by kevin-bates** [comment]
> Should we allow the user to discover + set these at all?

System-owned properties, in general, were brought up on the issue (https://github.com/elyra-ai/elyra/issues/1696#issuecomment-848259469) and we should probably discuss it as something outside of this particular PR.

Generally speaking, I don't think users should be allowed to _set_ system-owned envs.   `ELYRA_RUNTIME_ENV` is a system-owned env that the system controls, yet for use by pipeline authors.  It should not be user-settable.   However, when we considered accessing a Gateway server from a (non-local) runtime, we were going to expose all of the gateway configuration "knobs" via `ELYRA_GATEWAY_` variables.  So those _could_ be considered something that should be user-settable.  But are they considered _system-owned_ in that case?  I think so.  So then it really becomes an issue as to how the application would know what is and what is not user-settable. Since we don't have any `ELYRA_`-prefixed envs that we think should be user-settable at this moment, we could make a statement that `ELYRA_`-prefixed envs are system-owned and non-user-settable.  But, envs make great "knobs" and we might want a user-settable, system-owned `ELYRA`-prefixed env someday.  :smile:

So perhaps the answer is:  _`ELYRA_`-prefixed envs are system-owned, and, yes, you _can_ set them.  However, non-user-settable system-owned envs will NOT honor your value and WILL BE overridden by the server._ Because they are truly system-owned.  In this case, the best "discovery" mechanism for user's setting system-owned envs will be the documentation.

(We also have some `AWS_` variables that should be considered system-owned, despite the fact that they derive their _values_ from runtime-configuration properties (i.e., user-controlled metadata).  These should not be settable outside of the values the _server_ pulls from the runtime configuration to set them - nor should they appear anywhere (so non-discoverable).)

Just my two cents.

**Comment by ptitzler** [comment]
> we should probably discuss it as something outside of this particular PR.

I agree. Created https://github.com/elyra-ai/elyra/issues/1704 for follow-up

> the best "discovery" mechanism for user's setting system-owned envs will be the documentation.

Created https://github.com/elyra-ai/elyra/issues/1703 for follow-up.

File: elyra/pipeline/pipeline.py [hunk 1]
```diff
@@ -141,20 +141,19 @@ def env_vars_as_dict(self, logger: Optional[Logger] = None) -> Dict:
         """
         Operation stores environment variables in a list of name=value pairs, while
         subprocess.run() requires a dictionary - so we must convert.  If no envs are
-        configured on the Operation, the existing env is returned, otherwise envs
-        configured on the Operation are overlayed on the existing env.
+        configured on the Operation, an empty dictionary is return, otherwise envs
```
Comment (ptitzler) [inline elyra/pipeline/pipeline.py:144]: typo

```suggestion
        configured on the Operation, an empty dictionary is returned, otherwise envs
```

Comment (kevin-bates) [reply elyra/pipeline/pipeline.py]: Done.

File: elyra/pipeline/pipeline.py [hunk 2]
```diff
@@ -141,20 +141,19 @@ def env_vars_as_dict(self, logger: Optional[Logger] = None) -> Dict:
...
+        configured on the Operation, an empty dictionary is returned, otherwise envs
+        configured on the Operation are converted to dictionary entries and returned.
         """
         envs = {}
         for nv in self.env_vars:
             if nv and len(nv) > 0:
                 nv_pair = nv.split("=")
                 if len(nv_pair) == 2 and nv_pair[0].strip():
                     envs[nv_pair[0]] = nv_pair[1]
+                elif logger:
```
Comment (ptitzler) [inline elyra/pipeline/pipeline.py:153]: Follow up to this scenario:
 - User sets variable to `MYVAR=3=2`.
 - VPE does not raise a validation error
 - A warning message is logged and the variable ignored but the user never explicitly notified.

```
Could not process environment variable entry `MYVAR=3=2`, skipping...
```

Would it be better to 
- treat this as fatal error on the back-end
- add code to the validation engine to alert the user?

Comment (ptitzler) [reply elyra/pipeline/pipeline.py:153]: "add code to the validation engine to alert the user?" -> new issue

**Comment by ptitzler** [comment]
Tested local/kfp/airflow and `ELYRA_RUNTIME_ENV`. Seems to work now as expected.

File: elyra/pipeline/pipeline.py (hunk 2 above)
Comment (kevin-bates) [reply elyra/pipeline/pipeline.py:153]: I think we should preserve the intention of the user and not treat any inability to parse environment variables as fatal at this particular location.  Should we treat these as errors detected during validation?  Perhaps, but that's not in the scope of this issue.

Regarding `MYVAR=` scenarios.  This was discussed in today's scrum and the consensus was that this particular _non-value_ scenario should not result in `MYVAR`'s inclusion in the environment of the runtime.  However, an explicit, yet empty, scenario like `MYVAR=""` _should_ result in `MYVAR`'s inclusion in the environment.

If you're in agreement here, I can include this functionality in this PR - please advise.

Comment (ptitzler) [reply elyra/pipeline/pipeline.py:153]: > If you're in agreement here, 

Yes, that makes sense

How about handling multiple `=` signs in the backend code? https://github.com/elyra-ai/elyra/pull/1701#discussion_r639863979. Having thought about this some more I believe `A=2=3` could be a valid scenario and should not be ignored. I am thinking about the component library here, which uses environment variables extensively and might have a need to have an `=` sign somewhere in the value.

**Review by ptitzler — APPROVED** [review]
LGTM

**Review by lresende — APPROVED** [review]

**Pull request merged**